### ISSUE TITLE
app.emit if not a function

### ISSUE THREAD

**Issue by dottodot** [feathersjs/feathers#1183]
I have a strange problem that I can't get to the bottom off and it's only recently started happening and on 2 seperate sites, and only manifests after a period of time.

When the user goes to login I get the error `app.emit if not a function` (server error) and the login fails. I've not found a way to replicate it and the problem clears on a restart. 

I thought the issue might be something to do with feathers-sync but I've moved the initialisation on that to after authentication so that it doesn't run on those requests.

**Comment by daffl** [comment]
Have you tried disabling `feathers-sync` completely? Is it still happening then? It's always tricky with problems like this if they are not reproducible since anything could really be the cause.

**Comment by dottodot** [comment]
I've not as yet. One thing I've noticed that's odd is it's a 500 server error which I'm seeing in the client side logs but I'm not seeing a corresponding error in the server logs yet I can see errors such as BadRequest and NotFound

**Comment by dottodot** [comment]
OK I've disabled `feathers-sync` so should know pretty soon if it's to do with that as the issue seems to manifest with 6 hours of a restart.

**Comment by dottodot** [comment]
OK it's nothing to do with feathers-sync just happened again with sync disabled.

**Comment by daffl** [comment]
Do you have more of a stack trace or information about in which case this may happen? There are two cases where `app.emit` is being called in the framework

1. On `login` and `logout`
2. On `publish` when any event is published to its channels

In both cases I'm not seeing how it would be possible for `app` to be anything other than the app unless something else messes with it.

**Comment by dottodot** [comment]
That's the thing I'm a bit baffled about, how can my client be getting a 500 Server Error but I'm not getting a error In my Sentry logs for the server. All other errors seem to get logged.

The other thing I've noticed is if you try logging in a few times is can be successful on one of the attempts. Just wondering if it's something to do with it running on a Kubernetes cluster i.e the attempts are being load balanced to different servers, one works one doesn't.

**Comment by daffl** [comment]
That is sometimes possible when the error happens outside of the common flow. 
One case I realized where this actually could happen is when `req.app` is not set in the `emitEvents` middleware. I take it this happens during authentication?

**Comment by suitupalex** [comment]
Just want to mention I've been experiencing the exact same issue as well. Will update and see if I'm still getting the error.

**Comment by dottodot** [comment]
Don't seem to be getting the `app.emit` error anymore but just got a `this.app.get is not a function` error

**Comment by suitupalex** [comment]
@daffl I'm getting the same thing as @dottodot plus other weirdness.

If I'm using local auth, I either get:

```
Jan 27 12:35:36 agg-ms-api-production app[web] info Sun, 27 Jan 2019 18:35:36 GMT agg:v0:error TypeError: this.app.get is not a function
Jan 27 12:35:36 agg-ms-api-production app[web] info     at IncomingMessage.protocol (/app/node_modules/express/lib/request.js:310:24)
Jan 27 12:35:36 agg-ms-api-production app[web] info     at IncomingMessage.secure (/app/node_modules/express/lib/request.js:336:15)
Jan 27 12:35:36 agg-ms-api-production app[web] info     at /app/node_modules/@feathersjs/authentication/lib/express/set-cookie.js:36:18
Jan 27 12:35:36 agg-ms-api-production app[web] info     at WebFrameworkShim.applySegment (/app/node_modules/newrelic/lib/shim/shim.js:1386:20)
Jan 27 12:35:36 agg-ms-api-production app[web] info     at _applyRecorderSegment (/app/node_modules/newrelic/lib/shim/shim.js:977:20)
Jan 27 12:35:36 agg-ms-api-production app[web] info     at _doRecord (/app/node_modules/newrelic/lib/shim/shim.js:956:17)
Jan 27 12:35:36 agg-ms-api-production app[web] info     at /app/node_modules/newrelic/lib/shim/shim.js:927:24
Jan 27 12:35:36 agg-ms-api-production app[web] info     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
Jan 27 12:35:36 agg-ms-api-production app[web] info     at next (/app/node_modules/express/lib/router/route.js:137:13)
Jan 27 12:35:36 agg-ms-api-production app[web] info     at boundNext (/app/node_modules/newrelic/lib/shim/webframework-shim.js:999:30)
```

Which sometimes returns as `this.app.get is not a function` or `Invalid login` to the client.

If I try to hit an oauth 1 or 2 endpoint or any provider type (I use Discord, Facebook, Twitch, Twitter, and YouTube), I get `req.app.authenticate` is not a function with different stack lengths:

```
Jan 27 12:37:14 agg-ms-api-production app[web] info Sun, 27 Jan 2019 18:37:14 GMT agg:v0:error TypeError: req.app.authenticate is not a function
Jan 27 12:37:14 agg-ms-api-production app[web] info     at /app/node_modules/@feathersjs/authentication/lib/express/authenticate.js:24:13
Jan 27 12:37:14 agg-ms-api-production app[web] info     at WebFrameworkShim.applySegment (/app/node_modules/newrelic/lib/shim/shim.js:1386:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info Sun, 27 Jan 2019 18:37:14 GMT agg:v0:error TypeError: req.app.authenticate is not a function
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _applyRecorderSegment (/app/node_modules/newrelic/lib/shim/shim.js:977:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _doRecord (/app/node_modules/newrelic/lib/shim/shim.js:956:17)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at /app/node_modules/newrelic/lib/shim/shim.js:927:24
Jan 27 12:37:14 agg-ms-api-production app[web] info     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at next (/app/node_modules/express/lib/router/route.js:137:13)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at WebFrameworkShim.applySegment (/app/node_modules/newrelic/lib/shim/shim.js:1386:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _applyRecorderSegment (/app/node_modules/newrelic/lib/shim/shim.js:977:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info TypeError: req.app.authenticate is not a function
Jan 27 12:37:14 agg-ms-api-production app[web] info     at /app/node_modules/@feathersjs/authentication/lib/express/authenticate.js:24:13
Jan 27 12:37:14 agg-ms-api-production app[web] info     at WebFrameworkShim.applySegment (/app/node_modules/newrelic/lib/shim/shim.js:1386:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _applyRecorderSegment (/app/node_modules/newrelic/lib/shim/shim.js:977:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _doRecord (/app/node_modules/newrelic/lib/shim/shim.js:956:17)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at /app/node_modules/newrelic/lib/shim/shim.js:927:24
Jan 27 12:37:14 agg-ms-api-production app[web] info     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at next (/app/node_modules/express/lib/router/route.js:137:13)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at WebFrameworkShim.applySegment (/app/node_modules/newrelic/lib/shim/shim.js:1386:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _applyRecorderSegment (/app/node_modules/newrelic/lib/shim/shim.js:977:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at /app/node_modules/@feathersjs/authentication/lib/express/authenticate.js:24:13
Jan 27 12:37:14 agg-ms-api-production app[web] info     at WebFrameworkShim.applySegment (/app/node_modules/newrelic/lib/shim/shim.js:1386:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _applyRecorderSegment (/app/node_modules/newrelic/lib/shim/shim.js:977:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _doRecord (/app/node_modules/newrelic/lib/shim/shim.js:956:17)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at /app/node_modules/newrelic/lib/shim/shim.js:927:24
Jan 27 12:37:14 agg-ms-api-production app[web] info     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at next (/app/node_modules/express/lib/router/route.js:137:13)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at WebFrameworkShim.applySegment (/app/node_modules/newrelic/lib/shim/shim.js:1386:20)
Jan 27 12:37:14 agg-ms-api-production app[web] info     at _applyRecorderSegment (/app/node_modules/newrelic/lib/shim/shim.js:977:20)
```

Restarting fixes the problem temporarily, but it definitely resurfaces and breaks authentication for all of my users until the restart.

I tried rolling back several patch versions across all the auth libraries, but I'm still getting the issue (just with app.emit or app.authenticate)

**Comment by dottodot** [comment]
@daffl Do you have any suggestions as to how we can help find the cause? There must be something that's triggering the problem hence it's randomness. It also is likely to be a package that's been updated recently as I've only just started seeing the issue.

**Comment by daffl** [comment]
So where this is happening after the fix now is in https://github.com/feathersjs/feathers/blob/master/packages/authentication/lib/express/set-cookie.js#L36. It looks like `req.app` all of a sudden became something it is not supposed to. The strange part is that even if it was just a [normal Express application](https://expressjs.com/en/4x/api.html#req.app) `app.get` would still be available.

**Comment by dottodot** [comment]
This is looking for a needle in a haystack. Could be quite literally anything that's causing the issue and is impacting my users in a big way. Has anyone found a solution to this issue as I'm getting pretty desperate for a fix.

**Comment by dottodot** [comment]
@daffl I've found the issue. Think it's something to do with adding `.Model` 

```
async create(data, params) {
    const Discount = this.app.service('discount').Model;

    let discount_query = {
      $or: [
        {
          qty: undefined,
          expiry: {
            $gt: new Date()
          }
        },
        {
          qty: {
            $gt: 0
          },
          expiry: {
            $gt: new Date()
          }
        }
      ]
    };

    if (data.code) {
      discount_query.code = { $regex: new RegExp(data.code), $options: 'i' };
    }

    const discount = await Discount.findOne(discount_query);
    return discount;
  }
```

**Comment by dottodot** [comment]
Scrap that nothing to do with .Model it seems to be when throwing an error in a hook i.e

```
    throw new errors.NotAcceptable(
      'Code not found, please check and try again.'
    );
```

**Comment by dottodot** [comment]
OK really have found the issue now.

I had a custom transport in the winston logger that uses Sentry so it's something to do with that. Disabling the transport resolves the issue.

**Comment by dottodot** [comment]
Not sure if this is a Feathers or Sentry issue. Is the FeathersError meant to attach the hook to it which contains a reference to the app?

**Comment by dottodot** [comment]
It's a Sentry issue
https://github.com/getsentry/sentry-javascript/issues/1855
https://github.com/getsentry/sentry-javascript/issues/1859

**Comment by daffl** [comment]
Thank you for figuring this out and sharing it.

**Comment by kamilogorek** [comment]
Fixed in `4.5.4` – https://github.com/getsentry/sentry-javascript/releases/tag/4.5.4
Sorry for the inconvenience :)
